subprojects/core/src/main/java/org/gradle/api/internal/tasks/properties/ValidationActions.java

```diff
@@ -110,9 +116,81 @@ public void doValidate(String propertyName, Object values, TaskValidationContext
+                .addPossibleSolution(() -> "Make sure the " + lowerKind + " exists before the task is called")
+                .addPossibleSolution(() -> "Make sure that the task which produces the " + lowerKind + " is declared as an input")
+                .documentedAt("validation_problems", "input_file_does_not_exist");
+        });
+    }
+
+    private static void reportUnexpectedInputKind(TaskValidationContext context, String kind, String propertyName, File input) {
+        context.visitPropertyProblem(problem -> {
+            String lowerKind = kind.toLowerCase();
+            problem.withId(ValidationProblemId.UNEXPECTED_INPUT_FILE_TYPE)
+                .forProperty(propertyName)
+                .reportAs(ERROR)
+                .withDescription(() -> lowerKind + " '" + input + "' is not a " + lowerKind)
+                .happensBecause(() -> "Expected an input to be a " + lowerKind + " but it was a " + actualKindOf(input))
+                .addPossibleSolution(() -> "Use a " + lowerKind + " as an input")
+                .addPossibleSolution(() -> "Declare the input as a " + actualKindOf(input) + " instead")
+                .documentedAt("validation_problems", "unexpected_input_file_type");
+        });
+    }
+
+    private static void reportCannotWriteToDirectory(String propertyName, TaskValidationContext context, File directory, String cause) {
+        context.visitPropertyProblem(problem ->
+            problem.withId(ValidationProblemId.CANNOT_WRITE_OUTPUT)
+                .reportAs(ERROR)
+                .forProperty(propertyName)
+                .withDescription(() -> "is not writable because " + cause)
+                .happensBecause(() -> "Expected '" + directory + "' to be a directory but it's a " + actualKindOf(directory))
+                .addPossibleSolution("Make sure that the '" + propertyName + "' is configured to a directory")
+                .documentedAt("validation_problems", "cannot_write_output")
+        );
+    }
+
+
+    private static void reportCannotWriteToFile(String propertyName, TaskValidationContext context, String cause) {
+        context.visitPropertyProblem(problem ->
+            problem.withId(ValidationProblemId.CANNOT_WRITE_OUTPUT)
+                .reportAs(ERROR)
+                .forProperty(propertyName)
+                .withDescription(() -> "is not writable because " + cause)
+                .happensBecause(() -> "Cannot write a file to a location pointing at a directory")
+                .addPossibleSolution(() -> "Configure '" + propertyName + "' to point to a file, not a directory")
+                .documentedAt("validation_problems", "cannot_write_output")
+        );
+    }
+
+
+    private static String actualKindOf(File input) {
+        if (input.isFile()) {
+            return "file";
+        }
+        if (input.isDirectory()) {
+            return "directory";
+        }
+        return "unexpected file type";
```

Comment:
I suppose we should also check whether the file exists here, and then report `non-existing file` or something like that.

Reply:
That's not possible because we have other validation rules for non-existing files. It's really code in case something fishy happens. Symlinks are resolved, so we shouldn't reach this point. So I guess what we could have is file types which are neither files nor directories, nor symlinks. I'm not sure what actual things fall in that category.

Reply:
I was thinking about the case when we can't create the output location, because the parent file is an actual file. Anyway, let's leave it as is.